src/feat/online-feature.cc

```diff
@@ -24,18 +24,55 @@
 
 namespace kaldi {
 
+RecyclingVector::RecyclingVector(int items_to_hold) : items_to_hold_(items_to_hold),
+                                                  first_available_index_(0) { }
+
+RecyclingVector::~RecyclingVector() {
+  DeletePointers(&items_);
+}
+
+Vector<BaseFloat> *RecyclingVector::Retrieve(int index) const {
+  if (index < first_available_index_) {
+    KALDI_ERR << "Attempted to retrieve feature vector that was "
+                 "already removed by the RecyclingVector (index = " << index << "; "
+              << "first_available_index = " << first_available_index_ << "; "
+              << "size = " << Size() << ")";
+  }
+  // 'at' does size checking.
+  return items_.at(index - first_available_index_);
+}
+
+void RecyclingVector::Store(Vector<BaseFloat> *item) {
+  // Internally, we keep at most 2x requested items_to_hold_ and clean up
+  // when this number is reached
+  const int size_before_erase = items_.size();
```

Comment:
I'm assuming the concern here is the efficiency, as to avoid droping the first element and moving the entire rig towards the head of the vector on every (overflowing) insertion. If so, consider `std::deque` as the internal container instead of the `std::vector`. Presumably, a library-implemented container maintains a good balance between overhead and efficiency of `at()` access. 2x the requested size may not always be a good heuristic.

Reply:
I considered this before but wasn't sure. Since you're the second person to suggest it and it results in less code, I guess it's better though.